src/models/batch_auction_model.rs

```diff
@@ -160,6 +170,155 @@ impl SettledBatchAuctionModel {
             .flat_map(|u| &u.execution)
             .all(|u| u.exec_plan.is_some())
     }
+
+    pub fn price(&self, token: H160) -> Option<&U256> {
+        self.prices.get(&token)
+    }
+
+    pub fn tokens(&self) -> HashSet<H160> {
+        self.prices
+            .iter()
+            .map(|(token, _)| *token)
+            .collect::<HashSet<H160>>()
+    }
+
+    pub fn insert_new_price(
+        &mut self,
+        splitted_trade_amounts: &HashMap<(H160, H160), (U256, U256)>,
+        query: SwapQuery,
+        swap: SwapResponse,
+        tokens: &BTreeMap<H160, TokenInfoModel>,
+    ) -> Result<()> {
+        let src_token = query.sell_token;
+        let dest_token = query.buy_token;
+        let (sell_amount, buy_amount) = match (
+            splitted_trade_amounts.get(&(src_token, dest_token)),
+            splitted_trade_amounts.get(&(dest_token, src_token)),
+        ) {
+            (Some((_sell_amount, _)), Some((buy_amount, substracted_sell_amount))) => {
+                (*substracted_sell_amount, *buy_amount)
+            }
+            (Some((_, _)), None) => (U256::zero(), U256::zero()),
+            (None, Some((_, _))) => (U256::zero(), U256::zero()),
+            (None, None) => (U256::zero(), U256::zero()),
+        };
+        let (sell_amount, buy_amount) = (
+            sell_amount.checked_add(swap.sell_amount).unwrap(),
+            buy_amount.checked_add(swap.buy_amount).unwrap(),
+        );
+
+        match (
+            self.prices.clone().get(&src_token),
+            self.prices.clone().get(&dest_token),
+        ) {
+            (Some(_), Some(_)) => return Err(anyhow!("can't deal with such a ring")),
+            (Some(price_sell_token), None) => {
+                self.prices.insert(
+                    query.buy_token,
+                    price_sell_token
+                        .checked_mul(sell_amount)
+                        .unwrap()
+                        .checked_div(buy_amount)
+                        .unwrap(),
+                );
+            }
+            (None, Some(price_buy_token)) => {
+                self.prices.insert(
+                    query.sell_token,
+                    price_buy_token
+                        .checked_mul(buy_amount)
+                        .unwrap()
+                        .checked_div(sell_amount)
+                        .unwrap(),
+                );
+            }
+            (None, None) => {
+                if self.prices.is_empty() {
+                    self.prices.insert(
+                        query.sell_token,
+                        buy_amount.checked_mul(U256::from(SCALING_FACTOR)).unwrap(),
+                    );
+                    self.prices.insert(
+                        query.buy_token,
+                        sell_amount.checked_mul(U256::from(SCALING_FACTOR)).unwrap(),
+                    );
+                } else {
```

Comment:
Just want to understand what changes in this PR. Is it that in the case of already computed prices, you then use these as references to scale the prices of additional tokens accordingly?

Reply:
yes, exactly